mgr/dashboard: monitoring: refactor into ceph-mixin

MrFreezeex opened the pull request:
AFTER https://github.com/ceph/ceph/pull/43685

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests
- Teuthology
  - [ ] Completed teuthology run
  - [ ] No teuthology test necessary (e.g., documentation)

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

alfonsomthd commented:
@MrFreezeex Please add the tracker issue related to this PR.

alfonsomthd commented:
Hi @MrFreezeex ,
@epuertat and me areinviting you to join the next Ceph Dashboard community Daily Standup so you can introduce yourself :smiley:  and give us more insight on this PR.
The calendar meeting:
```
Ceph Dashboard Daily Standup
10:30 – 11:00 CET (Mondays, Tuesdays, Wednesdays, Fridays)
Google Meet Video call link: https://meet.google.com/bix-nypb-jma
```

MrFreezeex commented:
> Hi @MrFreezeex , @epuertat and me are inviting you to join the next Ceph Dashboard community Daily Standup so you can introduce yourself smiley and give us more insight on this PR.

Hi, sure I will try to join tomorrow!

MrFreezeex commented:
Apart from the cephadm e2e, the CI seems happy with my PR, I don't believe that this PR is the reason of this failure though.

If you review this PR some notes:
- The yaml of the alerts is different but it should have no semantic difference, I converted the original yaml to json then to jsonnet. The jsonnet file generated (`monitoring/ceph-mixin/alerts.libsonnet`) looks fine to me but feel free to take a look
- The jsonnet dashboards are now in `monitoring/ceph-mixin//dashboards/`, I splitted the original file (which generated all the dashboards) into the following files: `cephfs.libsonnet`,  `dashboards.libsonnet`, `general.libsonnet`,  `rbd.libsonnet`  and `rgw.libsonnet`. I didn't change much the content of the jsonnet besides splitting the file, putting all the dashboards in the `grafanaDashboards` dictionnary and some jsonnetfmt linting.
- The rest of it is about changing the dashboards and alerts generation into `monitoring/ceph-mixin` and changing this path everywhere else it is referenced
- This PR is about having a very basic conversion to mixin with some related directory change that looks somehow reasonable without too much refactoring the content of the jsonnet. I will most likely do some following PR to do more cleanup in the jsonnet dashboards and the grafana queries (and actually adding something in `config.libsonnet` ;)).

epuertat commented on monitoring/ceph-mixin/README.md:
```diff
@@ -0,0 +1,40 @@
+You can find all the generated files in the `output` directory and you can
+run `tox -egrafonnet-fix` to re-generate them.
+
+The ceph-mixin depends on some Jsonnet third party libraries located in the
+`vendor` directory. To update those libraries you will have to install
+[jsonnet-bundler](https://github.com/jsonnet-bundler/jsonnet-bundler) and run
+the command `jb update`.
```
We discussed the versioning of grafonnet files in the original PR. What's the reason not to download them at build time? Is it a best practice in the jsonnet/mixin world? If not, couldn't we instead add them as a `git subtree`?

MrFreezeex replied:
I didn't do it because this is managed by jsonnet-bundler and I thought that it would add another dependency. A `git subtree` would probably be in conflict with what jsonnet is doing :/.

I think we should choose between:
1. Having the final json and yaml for the dashboards and grafana in the repo with grafonnet
2. Having the final json and yaml for the dashboards and grafana in the repo without grafonnet BUT having jsonnet-bundler at least in the CI that check if the final json files are up to date
3. Having neither grafonnet nor the final yaml/json in the repo but add jsonnet and jsonnet-bundler as a build dependency

1) is convenient because you don't have to deal with dependencies
2) is what is done in master today but with a bit more complexity since we would have to install a go software not packaged for our distro (we could probably do a go install somewhere instead if we choose this)
3) need to have the dependencies installed at both build time and check time and is not super convenient for users that just want to download the json/yaml from the repo

I didn't see any project that keeps the vendor directory in their repo (I didn't see any that was directly in the build system and have the need to be packaged though). Some includes the final json/yaml directly in the repo some don't it depends.

MrFreezeex replied:
After thinking about this again, I would say adding go as a builld dependency and build jsonnet-bundler would be worth it because I want to add some other go program for the grafana dashboard (for instance this one https://github.com/grafana/dashboard-linter).

I don't know if we want to do the 3rd option (so about not adding the final json/yaml in the repo but building it). Maybe we want to wait for the missing dashboard which is not generated with jsonnet to be also generated with jsonnet before doing that (`ceph-cluster.json` IIRC).

epuertat replied:
@MrFreezeex thank you! Personally:
* I like having both the input-output files versioned, as that allows us to quickly detect how jsonnet/grafonnet changes impact on the generated JSONs.
* I'd prefer not to have the grafonnet code versioned, but if necessary, the `git tree`/`submodules` would be my preferred choices... :-/
* Regarding the build deps, sooner or later (hopefully sooner) we should end up building Ceph in a containerized pipeline, so we will be able to run some pipeline stages in the specific environments where the required tools are available (jsonnet, jsonnet-bundler, promtool, etc).

That said, I'm all for a pragmatic approach and you seem to be quite experienced in this stuff, so I'm ok with it! I'd also like to hear from other Ceph devels that reviewed [the original PR](https://github.com/ceph/ceph/pull/42194): @tchaikov @p-se @smithfarm @pcuzner @djgalloway @ktdreyer

MrFreezeex replied:
@epuertat I pushed two commits about that (the wip commit to be squashed ofc):
- https://github.com/ceph/ceph/pull/44059/commits/ad9ac597bf0682adf245305a7072a91aa6317d8b
- https://github.com/ceph/ceph/pull/44059/commits/d6beb370b2b2e8fddee37ae63c8a666af2c211ea

It adds go as a build dependency and install/run jsonnet bundler for make check. WDYT?

epuertat replied:
@MrFreezeex thanks a lot! Makes sense. I don't think installing `go` would be anything quirky in a project closely related to k8s.

smithfarm commented on monitoring/ceph-mixin/README.md:
```diff
@@ -0,0 +1,40 @@
+## Ceph mixin
```
I see several variations of the name:

1. `Ceph mixin`
2. `ceph-mixin`
3. `the ceph-mixin`

Which one is "the real McCoy"? I suggest to choose one variant and stick to it. Having multiple variants of a proper name is problematic! 

If it is a proper name, there should probably not be any "the" before it - unless of course the "the" is actually a part of the name (as in, e.g., "The Thing") . . . but in that case the "the" should appear *everywhere*.

If the name is `ceph-mixin` and it appears as the first word of a sentence, it should probably be capitalized.

MrFreezeex replied:
I no longer refer to the term `ceph-mixin`. I took some inspiration of the readme of [kubernetes-mixin](https://github.com/kubernetes-monitoring/kubernetes-mixin) and rook's [ceph-mixin](https://github.com/ceph/ceph-mixins).

smithfarm commented on monitoring/ceph-mixin/README.md:
```diff
@@ -0,0 +1,40 @@
+
+### Grafana dashboards for Ceph
+
+Here you can find a collection of [Grafana](https://grafana.com/grafana)
```
What does "here" mean? Where can the user find the dashboards, exactly?

And which user is being addressed here? A reader of a web page, or someone reading RPM documentation?

How do the dashboards get installed on a system?

MrFreezeex replied:
This is basically a copy/paste of the old `monitoring/grafana/dashboards` README. It probably makes more sense in the old path because there was the JSON dashboard directly available in the directory. I will try to write something that makes more sense in the current context.

Dashboards are installed with the package `ceph-grafana-dashboards` AFAIU but they can also be generated from jsonnet by a user. I will try to reflect that in the README as well :+1:.

MrFreezeex replied:
I rephrased some sentence it should be a bit better. I think this would be more addressed to a dev/someone digging up in the source code.

smithfarm commented on ceph.spec.in:
```diff
@@ -1450,7 +1450,7 @@ mkdir -p %{buildroot}%{_localstatedir}/lib/ceph/bootstrap-rbd
 
 # prometheus alerts
-install -m 644 -D monitoring/prometheus/alerts/ceph_default_alerts.yml %{buildroot}/etc/prometheus/ceph/ceph_default_alerts.yml
+install -m 644 -D monitoring/ceph-mixin/output/alerts/ceph_default_alerts.yml %{buildroot}/etc/prometheus/ceph/ceph_default_alerts.yml
```
As it stands, all the current `ceph.spec.in` changeset does is break the `ceph-prometheus-alerts` subpackage.

Further down in the spec file there is a section that looks like this:

```
%files prometheus-alerts
%if 0%{?suse_version}
%attr(0755,root,root) %dir %{_sysconfdir}/prometheus
%endif
%attr(0755,root,root) %dir %{_sysconfdir}/prometheus/ceph
%config %{_sysconfdir}/prometheus/ceph/ceph_default_alerts.yml
```

You'll need to make this change there, too.

MrFreezeex replied:
If I understand correctly I don't need to update this part of the spec file because I'm doing the install at the exact same place. But maybe I'm wrong?

MrFreezeex commented:
I changed the path of the generated files to be more in line with https://monitoring.mixins.dev/#generate-config-files. Also added a makefile the same reason and for convenience.

As discussed here https://github.com/ceph/ceph/pull/44059#discussion_r756742874, I added golang as a build dependency and I am building jsonnet-bundler (as it is pretty much not packaged anywhere). The PR should be ready again for further review!

epuertat commented on monitoring/ceph-mixin/alerts.libsonnet:
```diff
@@ -0,0 +1,875 @@
+{
+  prometheusAlerts+:: {
+    groups+: [
+      {
+        name: 'cluster health',
+        rules: [
+          {
+            alert: 'CephHealthError',
+            expr: 'ceph_health_status == 2',
```
Given that we're not using any jsonnet preprocessor magic here and that YAML is far more readable/user-friendly, why not passing through the original YAML, or if really needed, generating this with the [`std.parseYaml`](https://jsonnet.org/ref/stdlib.html#:~:text=yields%20%7B%20%22foo%22%3A%20%22bar%22%20%7D.-,std.parseYaml(str),-Available%20since%20version)?

I'd go for a jsonnet file here if we had something like Grafonnet, but tbh I find most of the info here cannot be 'compressed' much further:
```py
PrometheusAlerts().addGroup(
    Alert(
      'CephHealthError',
      'ceph_health_status == 2',
      ...
    )
```

MrFreezeex replied:
Ah I don't remember precisely, I probably missed the fact that `std.parseYaml` existed in the first hand and probably saw a bunch of projects doing a similar thing than what I attempted here.

I will try to use `std.parseYaml` and see how it goes, this seems at first glance a good thing to do!

epuertat replied:
Thanks! My first suggestion was to simply bypass the yaml file, as I don't think we were doing any preprocessing in there.

MrFreezeex replied:
Ok so now I remember why I didn't do that. parseYaml was not in any jsonnet release for the past year or so but this has changed because v0.18.0 was released 15 days ago. So fortunately for us we will be able to do that :D!

epuertat commented on monitoring/ceph-mixin/Makefile:
```diff
@@ -0,0 +1,27 @@
+	jb install
+
+dashboards_out: vendor
+	jsonnet -J vendor -m $@ dashboards.jsonnet
```
This same command (as well as the alerts one) is repeated in 3 different places: in this Makefile, in a shell script, and in a toxfile... I'd rather have a single source of truth for generating this. Is it viable?

epuertat replied:
If you want to allow users customize the build, tox also allows you to pass [`{posargs}`](https://tox.wiki/en/latest/example/general.html) (not sure if that was the issue you mentioned here).

epuertat commented on monitoring/ceph-mixin/README.md:
```diff
@@ -0,0 +1,50 @@
+
+#### Requirements
+
+- [Status Panel](https://grafana.com/plugins/vonage-status-panel) installed on
+  your Grafana instance
```
The `grafana-piechart-panel` was also required. Is it by default now included?

MrFreezeex replied:
Looking at this again apparently there is now a pie chart panel by default in grafana (see this commit https://github.com/grafana/piechart-panel/commit/b4719bfa0452cf22661d7262bfe736af157dfd1c#diff-81bc52c095e350118867727ac555c181ec0e4b8de06521f704677f7a60cdb5b8R5). Not sure if it is compatible directly though, I will check this!

epuertat replied:
Great! If we could get rid of the status panel by using some other built-in chart (any clue?), we could replace our custom Grafana image with the vanilla one.

It seems that there are a couple of built-in widgets that could potentially fit here (some of them have been introduced in Grafana 8.x):
* https://grafana.com/docs/grafana/latest/visualizations/status-history/
* https://grafana.com/docs/grafana/latest/visualizations/state-timeline/

MrFreezeex replied:
I think I will just add back the requirement in the README and check how we could use the builtin type in another PR.

epuertat commented on monitoring/ceph-mixin/alerts.libsonnet:
```diff
@@ -0,0 +1,875 @@
+            annotations: {
+              documentation: 'https://docs.ceph.com/en/latest/rados/operations/health-checks#mon-disk-crit',
+              summary: 'Disk space on at least one monitor is critically low',
+              description: "The free space available to a monitor's store is critically low (<5% by default).\nYou should increase the space available to the monitor(s). The\ndefault location for the store sits under /var/lib/ceph. Your monitor hosts are;\n{{- range query \"ceph_mon_metadata\"}}\n  - {{ .Labels.hostname }}\n{{- end }}\n",
```
One thing I love from jsonnet is verbatims (but YAML supports that too):
```suggestion
              description: |||
The free space available to a monitor's store is critically low (<5% by default).
You should increase the space available to the monitor(s). The
default location for the store sits under /var/lib/ceph. Your monitor hosts are;
{{- range query "ceph_mon_metadata"}}
  - {{ .Labels.hostname }}
{{- end }}
|||,
```

MrFreezeex replied:
I was also planning to do this kind of cleanup in a followup PR (same as https://github.com/ceph/ceph/pull/44059#discussion_r770604922)

epuertat replied:
Perfect! Yeah, given the size of this PR no need to make it heavier.

epuertat commented on monitoring/ceph-mixin/dashboards/cephfs.libsonnet:
```diff
@@ -0,0 +1,103 @@
+local u = import 'utils.libsonnet';
+
+{
+  grafanaDashboards+:: {
+    'cephfs-overview.json':
+      local CephfsOverviewGraphPanel(title, formatY1, labelY1, expr, legendFormat, x, y, w, h) =
+        u.graphPanelSchema({},
+                           title,
+                           '',
+                           'null',
+                           false,
+                           formatY1,
+                           'short',
+                           labelY1,
+                           null,
+                           0,
+                           1,
+                           '$datasource')
+        .addTargets(
+          [u.addTargetSchema(expr, 1, 'time_series', legendFormat)]
+        ) + { gridPos: { x: x, y: y, w: w, h: h } };
```
I really liked that you formatted all this code. Now it's easier to read.

epuertat commented on monitoring/ceph-mixin/dashboards_out/cephfs-overview.json:
```diff
@@ -64,6 +64,7 @@
          "datasource": "$datasource",
          "description": "",
          "fill": 1,
+         "fillGradient": 0,
```
Curious: was this due to some upgrade in grafonnet version?

MrFreezeex replied:
Yes exactly, since with this commit https://github.com/grafana/grafonnet-lib/commit/7ba2afd9f2f37a24178ee96a7d8944db2bd8db47 on grafonnet!

epuertat replied:
Thanks!

epuertat commented on monitoring/ceph-mixin/dashboards_out/hosts-overview.json:
```diff
@@ -514,7 +514,7 @@
          "targets": [
             {
-               "expr": "sum (\n\t(\n\t\tirate(node_network_receive_bytes{instance=~\"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*\",device!=\"lo\"}[1m]) or\n\t\tirate(node_network_receive_bytes_total{instance=~\"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*\",device!=\"lo\"}[1m])\n\t) unless on (device, instance)\n\tlabel_replace((bonding_slaves > 0), \"device\", \"$1\", \"master\", \"(.+)\")\n) + ...
+               "expr": "sum (\n        (\n                irate(node_network_receive_bytes{instance=~\"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*\",device!=\"lo\"}[1m]) or\n                irate(node_network_receive_bytes_total{instance=~\"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*\",device!=\"lo\"}[1m])\n        ) unless on (device, instance)\n        label_replace((bonding_slaves > 0), \"de ...
```
I understand it's not a pleasant work, but what about replacing these with verbatims?
```suggestion
               "expr": |||
sum (
        (
                irate(node_network_receive_bytes{instance=~"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*",device!="lo"}[1m]) or
                irate(node_network_receive_bytes_total{instance=~"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*",device!="lo"}[1m])
        ) unless on (device, instance)
        label_replace((bonding_slaves > 0), "device", "$1", "master", "(.+)")
) +
sum (
        (
                irate(node_network_transmit_bytes{instance=~"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*",device!="lo"}[1m]) or
                irate(node_network_transmit_bytes_total{instance=~"($osd_hosts|mon_hosts|mds_hosts|rgw_hosts).*",device!="lo"}[1m])
        ) unless on (device, instance)
        label_replace((bonding_slaves > 0), "device", "$1", "master", "(.+)")
        )
|||
```

MrFreezeex replied:
I was actually planning to do this in a followup PR (and other cleanups as well). I can do it in this PR if you think that's more appropriate though.

epuertat replied:
Yeah, it can wait for that follow-up!

epuertat commented on monitoring/ceph-mixin/tests_alerts/validate_rules.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python3 -u
+#!/usr/bin/env python3
```
Nice one! I'm afraid we're not consistent with this best practice.

MrFreezeex replied:
Yeah, it didn't work on my system that's why I did it actually :)! (the python of the virtual env generated by tox is not used with the previous shebang)

epuertat commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -153,7 +153,7 @@ commands =
 [testenv:check]
 commands =
-    python ci/check_grafana_dashboards.py frontend/src/app ../../../../monitoring/grafana/dashboards
+    python ci/check_grafana_dashboards.py frontend/src/app ../../../../monitoring/ceph-mixin/dashboards_out
```
For non cephadm users, there's a way for them to publish these Grafana JSONs to their Grafana instance:

https://github.com/ceph/ceph/blob/e911f41672fb9bafdb4b4b053998fb275c81f70d/src/pybind/mgr/dashboard/grafana.py#L103-L110

So it'd be nice to update that location too.

epuertat reviewed (changes requested):
I cannot but praise the amount of quality work you're contributing here @MrFreezeex ! I just left a few observations over there.

Additionally, you'll need to rebase in order to incorporate the following changes (some of them not yet merged):
* https://github.com/ceph/ceph/pull/43991 (Merged)
* https://github.com/ceph/ceph/pull/44294 (might be merged soon)
* https://github.com/ceph/ceph/pull/43685 (might be merged soon)

And please, do not despair. If you need help/guidance/support/whatever with this PR, please ping us. You know what time we have our daily standup ;-)

MrFreezeex commented on monitoring/ceph-mixin/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+      add_tox_test(jsonnet-lint TOX_ENVS jsonnet-lint)
+      add_tox_test(jsonnet-check TOX_ENVS jsonnet-check)
+      add_tox_test(alerts-check TOX_ENVS alerts-check)
+      #add_tox_test(promql-query-test TOX_ENVS promql-query-test)
```
This is also commented out in the code https://github.com/ceph/ceph/blob/070e54178adcf81c520252ebf558a111bda6da72/monitoring/prometheus/tests/CMakeLists.txt#L3. The test doesn't pass IIRC. We might want to fix it beforehand though!

epuertat replied:
I think that's fixed now: https://github.com/ceph/ceph/pull/44163

MrFreezeex replied:
I messed up on a path and it was failing apparently. It's fixed now! Not sure if I should keep the check only on arm though. It should pass on ARM AFAIU.

MrFreezeex commented on monitoring/ceph-mixin/Makefile:
```diff
@@ -0,0 +1,27 @@
+all: fmt generate lint test
+
+fmt:
+	./lint-jsonnet.sh -i
+
+generate: dashboards_out  prometheus_alerts.yaml
+
+vendor:
+	jb install
+
+dashboards_out: vendor
+	jsonnet -J vendor -m $@ dashboards.jsonnet
+
+prometheus_alerts.yaml: vendor
+	jsonnet -J vendor -S alerts.jsonnet -o $@
+
+lint: generate
+	tox -epromql-query-lint
+	tox -elint
+
+test: generate
+	tox -ejsonnet-check
+	#tox -epromql-query-test
+	tox -ealerts-check
+check: test
+
+.PHONY: all fmt generate lint test check
```
This makefile is only provided as a convenient things for dev to manipulate the mixin. There is some reference on make commands on https://monitoring.mixins.dev/ and generally the projects that I saw have some similar commands.

epuertat replied:
I'm not familiar with the monitoring mixins quirks, so I'll leave this up to you, but I'd encourage you ensure we have a single source of truth, so the other 2 files basically point to this one (I'd find kinda unusual having a `tox -e lint` invoking a `make lint`, but personally prefer to have a single definition of build recipes).

MrFreezeex commented:
> I cannot but praise the amount of quality work you're contributing here @MrFreezeex ! I just left a few observations over there.

Thanks that's greatly appreciated :D!

> Additionally, you'll need to rebase in order to incorporate the following changes (some of them not yet merged):
> 
> * #43991 (Merged)
> * #44294 (might be merged soon)
> * #43685 (might be merged soon)
> 
> And please, do not despair. If you need help/guidance/support/whatever with this PR, please ping us. You know what time we have our daily standup ;-)

Thanks I don't feel that I'm stuck on this PR and I do understand that a PR like that could also take some time to review, don't worry about it! It's also not blocking me on anything important as well. Feel free to ping me as well if you want some more explanations and you feel that it would help that I join the standup.

FYI I am taking a winter vacation for almost three weeks starting next week. I will try to work on this when I come back (rebasing, trying to address your comments). Feel free to prioritize others PR knowing that!

epuertat commented:
> FYI I am taking some winter holydays for almost three weeks starting next week. I will try to work on this when I come back (rebasing, trying to address your comments). Feel free to prioritize others PR knowing that!

Thanks for the heads-up! Enjoy your holidays! I'll be also taking some days off. We may catch up when we're back!

MrFreezeex commented:
Hi @epuertat,

When do you want to merge this? There is https://github.com/ceph/ceph/pull/44522 that would probably conflict not sure of the order you want between those two.

This PR should be fairly ready, I treated all your comments and rebased it. I didn't add back the arm exception introduced in https://github.com/ceph/ceph/pull/44163 though as the arm test does succeed but I can put it back if you prefer though.

Also the cephadm e2e test is failing, not sure if this is related though...

MrFreezeex commented:
jenkins test api

MrFreezeex commented:
jenkins test dashboard

MrFreezeex commented:
jenkins test dashboard cephadm

epuertat commented:
> Hi @epuertat,
> 
> When do you want to merge this? There is #44522 that would probably conflict not sure of the order you want between those two.
> 
> This PR should be fairly ready, I treated all your comments and rebased it. I didn't add back the arm exception introduced in #44163 though as the arm test does succeed but I can put it back if you prefer.
> 
> Also the cephadm e2e test is failing, not sure if this is related though...

Hi @MrFreezeex,

I'd personally like to see this landing in a Quincy release soon, not sure if in the first/ongoing one (since this brings many file changes), or a later one.

In any case, we would need at least 2 approvals here (I have to re-review it, but you may almost take mine for granted!). Perhaps another review from @pcuzner or @smithfarm?

I made explicit the dependency on https://github.com/ceph/ceph/pull/44522 on this one, so we don't merge any other Grafana PR before this one.

Additionally, @pcuzner was suggesting the idea (short/mid-term?) of moving the `/monitoring` directory to the (already existing) [ceph-mixins](https://github.com/ceph/ceph-mixins) repo, so to establish a common monitoring repository for cephadm + Rook users.

MrFreezeex commented:
> I'd personally like to see this landing in a Quincy release soon, not sure if in the first/ongoing one (since this brings many file changes), or a later one.

It depends how far you want to backport dashboards/alerts related PR. I personally think it would be helpful to eventually backport this to pacific as this change is also about operator importing the code into their prometheus/grafana stack managed with jsonnet.

> Additionally, @pcuzner was suggesting the idea (short/mid-term?) of moving the `/monitoring` directory to the (already existing) [ceph-mixins](https://github.com/ceph/ceph-mixins) repo, so to establish a common monitoring repository for cephadm + Rook users.

Yes having a lighter git repo dedicated for that might be nice. We would probably need to have several mixins though because the rook repo has several kubernetes/rook focused alerts which should not be in the ceph one.

So if you want to move out this code of ceph/ceph you could have a single repo with two sub directories, one for the general ceph mixin and one for rook (like grafana does here https://github.com/grafana/jsonnet-libs), or two distinct repositories. Also it would be helpful to have a branch for each stable branch and backport things like the ceph repo is doing (and not like the rook ceph-mixins repo). I don't have any strong opinion on which approach is better, it is probably more an organization question than a technical one...

If you chose to move out the code, we would need a minimal jsonnet code here to render the dashboards/alerts and import the jsonnet code from the other repo and run `jb update` from time to time to update the ceph mixin. Most likely mgr/dashboards focused test would run on those `jb update` PR, so any interaction with mgr/dashboard and the json dashboards would be tested "late". I don't think it's that much worrying though because I don't expect too much issues but I could be wrong...

The rook mixin would probably also benefit from some ceph alerts currently defined in this repo and it should be really simple to do that with the move to jsonnet and this refactoring. To be clear this does not require to move the code to another repository and they can probably start doing that after this PR has been merged (and probably also after a few change here as well).

epuertat commented:
jenkins test make check

epuertat commented:
The other 2 failures seem unrelated to this

epuertat commented:
Since this PR modifies build files (hence the build/ops label), I'll launch a Shaman package build to ensure the changes don't break in any supported distro.

[[Packages]](https://shaman.ceph.com/builds/ceph/wip-epuertat-testing-2022-01-21-2006/)

epuertat commented:
> > I'd personally like to see this landing in a Quincy release soon, not sure if in the first/ongoing one (since this brings many file changes), or a later one.
> 
> It depends how far you want to backport dashboards/alerts related PR. I personally think it would be helpful to eventually backport this to pacific as this change is also about ceph users importing the code into their prometheus/grafana stack managed with jsonnet.

Yes, you're right! Pacific would be a decent target for this, since it's still at mid-lifecycle.

> > Additionally, @pcuzner was suggesting the idea (short/mid-term?) of moving the `/monitoring` directory to the (already existing) [ceph-mixins](https://github.com/ceph/ceph-mixins) repo, so to establish a common monitoring repository for cephadm + Rook users.
> 
> Yes having a lighter git repo dedicated for that might be nice. We would probably need to have several mixins though because the rook repo has several kubernetes/rook focused alerts which should not be in the ceph one.

Good point! It seems that jsonnet supports [parametrized outputs](https://jsonnet.org/learning/tutorial.html#parameterize-entire-config), so both flavors could still derive from the same codebase.

> So if you want to move out this code of ceph/ceph you could have a single repo with two sub directories, one for the general ceph mixin and one for rook (like grafana does here [grafana/jsonnet-libs](https://github.com/grafana/jsonnet-libs)), or two distinct repositories. I don't have any strong opinion on which approach is better, it is probably more an organization question than a technical one... Also it would be helpful to have a branch for each stable release of ceph and backport things like in ceph/ceph (and not like the rook ceph-mixins repo).
> 
> If you chose to move out the code, we would need a minimal jsonnet code in ceph/ceph to render the dashboards/alerts and import the jsonnet code from the other repo and run `jb update` from time to time to update the ceph mixin. Most likely mgr/dashboard focused test would run on those `jb update` PR, so any interaction with mgr/dashboard and the json dashboards would be tested "late". I don't think it's that much worrying though because I don't expect too much issues but I could be wrong...
> 
> The rook mixin would probably also benefit from some ceph alerts currently defined in ceph/ceph and it should be really simple to do that with the move to jsonnet and this refactoring (and maybe a few followup cleanups). To be clear this does not require to move the code to another repository.

I'm not getting why we would still need some jsonnet stuff in the main ceph/ceph repo... If we move everything under `/monitoring` to a separate repo (and fix the existing in-repo dependencies), both things could be completely standalone. The only thing left would be to decide how to release the artifacts (currently both Grafana Dashboards and Prometheus Alerts are released in 2 different packages).

MrFreezeex commented:
> Good point! It seems that jsonnet supports [parametrized outputs](https://jsonnet.org/learning/tutorial.html#parameterize-entire-config), so both flavors could still derive from the same codebase.

You could theorically add something in the mixin config  (like here https://monitoring.mixins.dev/#customising-the-mixin) which is the way to do this with mixins. I would be still in favor of having two separate mixin with the rook mixin importing the ceph one. It would be more straightforward/clean IMO.

> I'm not getting why we would still need some jsonnet stuff in the main ceph/ceph repo... If we move everything under `/monitoring` to a separate repo (and fix the existing in-repo dependencies), both things could be completely standalone. The only thing left would be to decide how to release the artifacts (currently both Grafana Dashboards and Prometheus Alerts are released in 2 different packages).

Hmm yes you could do that as well. The jsonnet code will have a config at some point which would change the output, assuming that the config used by ceph is the default one it should be ok to do it thios way.

What I had in mind is to do a `import ceph-mixin/[...].libsonnet` and then ~20 lines of jsonnet code to output this into JSON in ceph/ceph essentially.

MrFreezeex commented:
> Since this PR modifies build files (hence the build/ops label), I'll launch a Shaman package build to ensure the changes don't break in any supported distro.
> 
> [[Packages]](https://shaman.ceph.com/builds/ceph/wip-epuertat-testing-2022-01-21-2006/)

It fails to clone jsonnet-bundler apparently, I will check on Monday what is happening exactly.

MrFreezeex commented:
Package builds now works: https://shaman.ceph.com/builds/ceph/wip-mrfreezeex-monit-mixin/d21700f24a1218020be5217d0a65fb6ad1349f96/ with https://github.com/ceph/ceph/pull/44059/commits/d21700f24a1218020be5217d0a65fb6ad1349f96.

epuertat commented:
jenkins test api

epuertat commented:
> Package builds now works: [shaman.ceph.com/builds/ceph/wip-mrfreezeex-monit-mixin/d21700f24a1218020be5217d0a65fb6ad1349f96](https://shaman.ceph.com/builds/ceph/wip-mrfreezeex-monit-mixin/d21700f24a1218020be5217d0a65fb6ad1349f96/) with [d21700f](https://github.com/ceph/ceph/commit/d21700f24a1218020be5217d0a65fb6ad1349f96).

Looks good. The [arm64 failure](https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/60056/consoleFull#-85362248744e9240e-b50a-4693-bac0-8a991bac86ac) seems unrelated to this:
> msg="container_linux.go:367: starting container process caused: error adding seccomp filter rule for syscall bdflush: permission denied"

I just retriggered the API tests.

epuertat reviewed (approved):
LGTM! Thank you @MrFreezeex ! I'll ask @ceph/dashboard folks tomorrow to review this (as well as @pcuzner ).

epuertat commented:
jenkins test dashboard

epuertat commented:
jenkins test dashboard cephadm

epuertat commented:
jenkins test dashboard

epuertat commented:
jenkins test dashboard cephadm

epuertat commented:
I also launched a [teuth run](https://pulpito.ceph.com/epuertat-2022-02-02_12:00:39-rados:dashboard-wip-mrfreezeex-monit-mixin-distro-basic-smithi/) to ensure this doesn't break anything during install.

djgalloway commented:
I think maybe this isn't working? https://jenkins.ceph.com/job/ceph-pull-requests/93945/

```
214/277 Test #264: run-tox-jsonnet-lint ......................***Failed    4.78 sec
Requirement already satisfied: tox in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (3.25.0)
Requirement already satisfied: pluggy>=0.12.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (1.0.0)
Requirement already satisfied: py>=1.4.17 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (1.10.0)
Requirement already satisfied: virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (20.14.1)
Requirement already satisfied: filelock>=3.0.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (3.6.0)
Requirement already satisfied: toml>=0.9.4 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (0.10.2)
Requirement already satisfied: six>=1.14.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (1.16.0)
Requirement already satisfied: packaging>=14 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (21.3)
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from packaging>=14->tox) (2.4.7)
Requirement already satisfied: platformdirs<3,>=2 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0->tox) (2.5.2)
Requirement already satisfied: distlib<1,>=0.3.1 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0->tox) (0.3.4)
jsonnet-lint create: /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/.tox/jsonnet-lint
jsonnet-lint installdeps: -rrequirements-grafonnet.txt
WARNING: Discarding $PYTHONPATH from environment, to override specify PYTHONPATH in 'passenv' in your configuration.
jsonnet-lint installed: jsondiff==2.0.0
jsonnet-lint run-test-pre: PYTHONHASHSEED='3895844870'
jsonnet-lint run-test: commands[0] | ./lint-jsonnet.sh --test
/home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/lint-jsonnet.sh: 5: jsonnetfmt: not found
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/lint-jsonnet.sh --test (exited with code 127)
```

MrFreezeex commented:
> I think maybe this isn't working? https://jenkins.ceph.com/job/ceph-pull-requests/93945/

Hmmm well the jsonnet-build tasks  should install jsonnetfmt (https://github.com/ceph/ceph/blob/master/monitoring/ceph-mixin/CMakeLists.txt#L14-L15 and https://github.com/ceph/ceph/blob/master/monitoring/ceph-mixin/jsonnet-build.sh), and I can see the jsonnet-build task being executed, so not sure why it doesn't find the binary... It's weird because the jsonnet binary seems to be installed as the jsonnet-check passed...

djgalloway commented:
Could `WARNING: Discarding $PYTHONPATH from environment, to override specify PYTHONPATH in 'passenv' in your configuration.` have anything to do with it?

MrFreezeex commented:
Hmmm I think this line of log is normal... Is there any sign to consistently reproduce the failure? Like perhaps using worker installed recently?

epuertat commented:
@djgalloway we should definitely try to move to containerized builds sooner than later... Having reproducible builds that don't mess with the system's deps is going to save us lots of environment-related pains.

djgalloway commented:
> @djgalloway we should definitely try to move to containerized builds sooner than later... Having reproducible builds that don't mess with the system's deps is going to save us lots of environment-related pains.

Do you mean building RPMs inside a container?  Let's discuss in the next CLT call.

epuertat commented:
> > @djgalloway we should definitely try to move to containerized builds sooner than later... Having reproducible builds that don't mess with the system's deps is going to save us lots of environment-related pains.
> 
> Do you mean building RPMs inside a container? Let's discuss in the next CLT call.

Sure (not necessarily RPMs, just the binaries).

tchaikov commented:
> I think maybe this isn't working? https://jenkins.ceph.com/job/ceph-pull-requests/93945/
> 
> ```
> 214/277 Test #264: run-tox-jsonnet-lint ......................***Failed    4.78 sec
> Requirement already satisfied: tox in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (3.25.0)
> Requirement already satisfied: pluggy>=0.12.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (1.0.0)
> Requirement already satisfied: py>=1.4.17 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (1.10.0)
> Requirement already satisfied: virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (20.14.1)
> Requirement already satisfied: filelock>=3.0.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (3.6.0)
> Requirement already satisfied: toml>=0.9.4 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (0.10.2)
> Requirement already satisfied: six>=1.14.0 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (1.16.0)
> Requirement already satisfied: packaging>=14 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from tox) (21.3)
> Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from packaging>=14->tox) (2.4.7)
> Requirement already satisfied: platformdirs<3,>=2 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0->tox) (2.5.2)
> Requirement already satisfied: distlib<1,>=0.3.1 in /home/jenkins-build/build/workspace/ceph-pull-requests/build/jsonnet-lint-virtualenv/lib/python3.8/site-packages (from virtualenv!=20.0.0,!=20.0.1,!=20.0.2,!=20.0.3,!=20.0.4,!=20.0.5,!=20.0.6,!=20.0.7,>=16.0.0->tox) (0.3.4)
> jsonnet-lint create: /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/.tox/jsonnet-lint
> jsonnet-lint installdeps: -rrequirements-grafonnet.txt
> WARNING: Discarding $PYTHONPATH from environment, to override specify PYTHONPATH in 'passenv' in your configuration.
> jsonnet-lint installed: jsondiff==2.0.0
> jsonnet-lint run-test-pre: PYTHONHASHSEED='3895844870'
> jsonnet-lint run-test: commands[0] | ./lint-jsonnet.sh --test
> /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/lint-jsonnet.sh: 5: jsonnetfmt: not found
> ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/lint-jsonnet.sh --test (exited with code 127)
> ```

this indeed does not work. another occurrence is https://jenkins.ceph.com/job/ceph-pull-requests/101536/console 

i created #47617 to address the test failure.
